### ISSUE TITLE
RFC: composable flux_future_t

### ISSUE THREAD

**Issue by grondo** [flux-framework/flux-core#1550]
I thought there was an existing issue open on the creation of composite `flux_future_t` objects,  but a quick search didn't turn anything up. If I missed an open issue, please let me know and I'll copy the ideas below into that issue and close this one.

Basic ability to create composite futures was implemented in #1188, however, the interface still has no inherent support for composability, and therefore composite futures have to be hand-coded, which is cumbersome and possibly error-prone.

This is a *very* preliminary idea of how we might implement composable futures for Flux. It is largely taken from existing work in C++ standard, e.g. see [N3857: Improvements to `std::future<T>`](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n3857.pdf)

One simple approach for composing futures might be to turn `flux_future_t` itself into a composite type. That is, a `flux_future_t` is always a composite of one *or more* `flux_future_t` future results, and the `then` continuation is executed in one of two cases:

 * "when all":  when all child `flux_future_t` results have been fulfilled.
 * "when any": when any one child `flux_future_t` result has been fulfilled.

In the common case, there is only one "child" `flux_future_t` and things work as they do now. In this case the "parent" and "child" future are equivalent and thus once the child future is fulfilled so is the parent.

In the composite case, each child's `then` continuation callback could check to see if their fulfillment completes the parent future. On completion of a composite parent, a new "completion" callback would have to be called to allow the implementation to gather possibly multiple results from all children futures and properly fulfill the parent. The "completion" callback would be similar to the current "initialization" callback and would be defined by the implementer at the time the composite is created.

The real trick here would be defining a nice interface for creating the `wait_all` and `wait_any` futures.
C doesn't have any nice tools like C++ for doing this, so the result will be much more verbose. The "completion" callback will also need access to all children futures, and will need to be able to distinguish which futures are which to be able to call the correct `_get()` interface for each.

Probably, the most straightforward approach would be to have functions that `push()` futures onto a composite, which was already created as a `when_any` or `when_all` variant, e.g. something like this (pseudocode)

Each child future is named so that the completion callback can identify which future corresponds to which operation.

```C
typedef void (*flux_future_completion_f) (flux_future_t *f, void *arg);
/*
 *  Create a composite future that will execute completion callback `cb`
 *   once *all* children futures have been fulfilled.
 *  The completion callback function should call flux_future_fulfill()
 */
flux_future_t * flux_future_wait_all (flux_future_completion_f cb, void *arg);

/*  Create a composite future that will execute completion callback `cb`
 *   once any one of the contained children futures have been fulfilled.
 *  The completion callback function should call flux_future_fulfill()
 */
flux_future_t * flux_future_wait_any (flux_future_completion_f cb, void *arg);

/*
 *  Push a child future `f` onto composite future `composite` with name `name`
 */
int flux_future_push (flux_future_t *composite,
                      flux_future_t *f, const char *name);


/*  Iterate over children of future `f` by name. If child is NULL return
 *   the first child. Returns NULL at end of list.
 */
const char * flux_future_next_child (flux_future_t *f, const char *child);
flux_future_t *flux_future_get_child (flux_future_t *f, const char *name);

/* Non-blocking check to determine if future `f` has been fulfilled */
bool flux_future_is_ready (flux_future_t *f);
```

I'll work on an example this afternoon to see if this interface makes any sense.

A final use case not covered by the above are chained futures -- where the initialization of one `flux_future_t` requires the fulfillment of another (for example, the result of one RPC is used in the creation of another). For this case, a `.next` future continuation is proposed in [N3865: More improvements for `std::future<T>`](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n3865.pdf), which allows a future to be declared as a continuation of another. I'm not sure if `flux_future_next` is the right method name, but something like that might be handy in the Flux future implementation:

```C 
/*  Return a future that is a continuation of existing future `f`. The initialization 
 *  function `init` will be called on fulfillment of `f`. The ancestor of the returned
 *  future can be accessed in the initialization function with `flux_future_get_prev (f)`
 */
flux_future_t *flux_future_next (flux_future_t *f, flux_future_init_f init, void *arg);

/*  Return any antecedent future for `f`. NULL if this future was not create with 
*   flux_future_next()
*/
flux_future_t *flux_future_get_prev (flux_future_t *f);
```

Any thoughts?

**Comment by grondo** [comment]
After chatting with @garlick I'd like to strike the last suggestion for chained futures, it will not work for any of the common cases such as running a series of chained RPCs which are dependent on one another (`flux_rpc` creates the future, so something like `flux_future_next` can't be used)

**Comment by grondo** [comment]
Additionally we discussed whether the "completion callback" (`flux_future_completion_f`) is required.

If we remove this callback, then composite futures would be automatically fulfilled by the implementation, and any post-processing or work required to gather results would be in a custom `flux_future_get_X()` call, which could traverse child futures (possibly recursively) to collect results.

This would imply that child futures cannot be destroyed until destruction of the parent.

The one case I'm still confused about is with `flux_future_wait_any()`. A common use case for this interface might be to add a timeout to an existing future. Since the implementation has no control over when the user calls `get()`, no post-processing such as cancelling outstanding timer on fulfillment of the existing future, or vice versa destroying the pending future from the timeout's `then` callback is possible. Therefore at the time the user calls `flux_future_get` the results may be ambiguous.. I'm probably just missing something obvious here though.

**Comment by grondo** [comment]
This was largely addressed in #1553.
